mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                        } else {
+                            results.put(we.getIndex(),
+                                        Error.get("update", MongoCrudConstants.ERR_SAVE_ERROR, we.getMessage()));
+                        }
+                    }
+                }
+                findConcurrentModifications(results);
+            }
+        }
+        reset();
+        return results;
+    }
+
+    /**
+     * This executes a query to find out documents with concurrent modification errors
+     */
+    protected void findConcurrentModifications(Map<Integer,Error> results) {
+        List<Object> updatedIds=new ArrayList<>(idsInBatch.size());
+        // Collect all ids without errors
+        int index=0;
+        for(Object id:idsInBatch) {
+            if(!results.containsKey(index))
+                updatedIds.add(id);
+            index++;
+        }
+        LOGGER.debug("checking for concurrent modifications:{}",updatedIds);
+        if(!updatedIds.isEmpty()) {
+            // documents with the given _ids and whose docver contains our docVer are the ones we managed to update
+            // others are failures
+            BasicDBObject query=new BasicDBObject(DOCVER_FLD,docVer);
+            query.append("_id",new BasicDBObject("$in",updatedIds));
+            DBCursor cursor=null;
+            try {
+                Set<Object> successfulIds=new HashSet<>(updatedIds.size());
+                cursor=collection.find(query,new BasicDBObject("_id",1));
+                while(cursor.hasNext()) {
+                    DBObject obj=cursor.next();
+                    Object id=obj.get("_id");
+                    successfulIds.add(id);
+                }
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
+                        }
+                    }
+                    index++;
+                }
+            } finally {
+                if(cursor!=null)
+                    cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Writes a replace query, which is:
+     * <pre>
+     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
+     * </pre>
+     */
+    private DBObject writeReplaceQuery(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        ObjectId top=null;
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null&&!list.isEmpty())
+                top=list.get(0);
+        }
+        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
+        query.append(DOCVER_FLD0,top);
```

Comment:
Sorry, I added an edit, not sure if you saw by email. Mainly it would be because, if it used often, then because both field are in index, it would be a "covered query" which are very efficient: https://docs.mongodb.com/manual/indexes/#covered-queries